### ISSUE TITLE
compactor: support arbitrary downsample intervals?

### ISSUE THREAD

**Issue by belm0** [thanos-io/thanos#2408]
The prometheus scrape rate in our deployment is 60s, so the hard-coded compactor intervals only provide 5x and 60x query speedup.

I would probably use 10m and 2h for this base interval if compactor was flexible.

**Comment by bwplotka** [comment]
Thanks! 

My question is.. why? (: Did you perform benchmarks showing that it will be faster/better in some way with those? We created downsampling for the generic use cases. Probably, for now, you have only scrape the 60s but maybe at some point you will add more targets with different intervals, right? (: Anyway current downsampling Lvl will work just fine with 1m interval.

Having this as dynamic parameters increase complexity for users extremely: 

* If you change such parameter what do you expect? Compactor rewriting all the thousands of 300GB blocks of data to change downsampling? What if you change to lower number, then that it's not possible without raw data (if you choose to have shorter retention of raw data).
* Queries and dashboards are built with the knowledge of underlying downsampling. It matters as otherwise `rate[lowerThanResolution]` will give you only one sample = no graph. By making it super dynamic you just know nothing at the query level.
* Users will start to want certain downsampling level just for single scrape target. 

To sum up: If we want to add this we need a very good reason and proper design that proves that this is worth for user. This is one of those features that it's easy to implement, but it will just explode in support questions and ambiguity of project usage (: But I am happy to discuss. :+1: If you agree with me, let's close this ticket.

**Comment by belm0** [comment]
Thank you-- those are fair points, especially about complexity of secondary consequences.

99% of our data is scraped at 60s, and I don't expect it to change, so the motivation is faster query times.  For a typical 15s scrape, 5m provides 20x less data, so we are looking for a similar factor at the 1st tier.

Another approach avoiding many of the issues is to consider a 3rd hard-coded downsampling level (e.g. 12h).  So at least for long time ranges our data reduction will be on par with configurations using 15s scrape.

**Comment by bwplotka** [comment]
I don't get why scrape interval really matters for next downsampling level. From `5m` to `1h` resolution it's exactly the same improvement no matter what initial scrape interval was (: 

`12h` is interesting idea. It makes resolution too high I think. For one year it's only 730 samples. Thanos can handle millions of samples, so it really might be unnecessary [unless you are from Asgard](https://scifi.stackexchange.com/questions/16783/how-long-do-asgardians-live-for) =DDDD

**Comment by belm0** [comment]
Downsampling is not merely one of the killer features of Thanos, but it's essential for Thanos to compete with prometheus-server in query execution performance.

For queries dealing with a lot of data, Thanos has a significant handicap vs. the locality of prometheus-server due to marshal/unmarshal overhead and shipping data across machines.  At least for such queries having a long timespan, downsampling lets Thanos meet or exceed Prometheus performance.

But let's review Thanos' relative performance boost for a user with 15s scrape vs 60s:

|case | samples (15s / 60s scrape) | vs. raw (15s / 60s scrape) |
|-|-|-|
|7-day query @ raw| 40,320 / 10,080| 1x|
|7-day query @ 5m| 2,016 | 20x / 5x|
|90-day query @ raw| 518,400 / 129,600 | 1x|
|90-day query @ 5m | 25,920 | 20x / 5x|
|90-day query @ 1h | 2,160 | 240x / 60x|

Since the user with 60s scrape is only getting a 5x or 60x boost vs. raw, especially in the 5m case it's not enough to counteract Thanos weaknesses in query performance.  He'll observe that Thanos query times are significantly higher than what he had on prometheus before migration.  That's what we're actually observing (and I've had a sanity-check thread with Frederic about it).

And the issue I reported in #2413 (migrated compacted blocks are not getting downsampled) makes this more painful.  We've migrated to Thanos, and expect the performance of queries with long time ranges to be on par with Prometheus-- but in fact they're much slower because 1h resolution is not available.
